anndata/tests/test_concatenate_disk.py

```diff
@@ -0,0 +1,853 @@
+    return request.param
+
+
+@pytest.fixture(params=["inner", "outer"])
+def join_type(request):
+    return request.param
+
+
+@pytest.fixture(params=[0, np.nan, np.pi])
+def fill_val(request):
+    return request.param
+
+
+@pytest.fixture(params=[0, 1])
+def axis(request):
+    return request.param
+
+
+@pytest.fixture(params=list(merge.MERGE_STRATEGIES.keys()))
+def merge_strategy(request):
+    return request.param
+
+
+def fix_known_differences(orig, result, backwards_compat=True):
+    """
+    Helper function for reducing anndata's to only the elements we expect to be
+    equivalent after concatenation.
+
+    Only for the case where orig is the ground truth result of what concatenation should be.
+
+    If backwards_compat, checks against what `AnnData.concatenate` could do. Otherwise checks for `concat`.
+    """
+    orig = orig.copy()
+    result = result.copy()
+
+    result.strings_to_categoricals()  # Should this be implicit in concatenation?
+
+    # TODO
+    # * merge varm, varp similar to uns
+    # * merge obsp, but some information should be lost
+    del orig.obsp  # TODO
+
+    if backwards_compat:
+        del orig.varm
+        del orig.varp
+        result.obs.drop(columns=["batch"], inplace=True)
+
+    # Possibly need to fix this, ordered categoricals lose orderedness
+    for k, dtype in orig.obs.dtypes.items():
+        if is_categorical_dtype(dtype) and dtype.ordered:
+            result.obs[k] = result.obs[k].astype(dtype)
+
+    return orig, result
+
+
+def adatas_to_zarr_paths(adatas, tmp_path):
+    """
+    Gets list of adatas, writes them and returns their paths as zarr
+    """
+    paths = None
+    if isinstance(adatas, typing.Mapping):
+        paths = {}
+        for k, v in adatas.items():
+            p = tmp_path / f"{k}.zarr"
+            v.write_zarr(p)
+            paths[k] = p
+    else:
+        paths = []
+        for i, a in enumerate(adatas):
+            p = tmp_path / f"{i}.zarr"
+            a.write_zarr(p)
+            paths += [p]
+    return paths
+
+
+# def test_concat_interface_errors_disk(tmp_path):
+#     adatas = [gen_adata((5, 10)), gen_adata((5, 10))]
+#     paths = adatas_to_zarr_paths(adatas, tmp_path)
+
+#     with pytest.raises(ValueError):
+#         concat_on_disk(paths, out_file=tmp_path / "test", axis=3)
+#     with pytest.raises(ValueError):
+#         concat_on_disk(paths, out_file=tmp_path / "test", join="not implemented")
+#     with pytest.raises(ValueError):
+#         concat_on_disk([], tmp_path / "test")
+
+
+def assert_eq_concat_on_disk(adatas, tmp_path, *args, **kwargs):
+    # create one from the concat function
+    res1 = concat(adatas, *args, **kwargs)
+    # create one from the on disk concat function
+    paths = adatas_to_zarr_paths(adatas, tmp_path)
+    out_name = tmp_path / "out.zarr"
+    concat_on_disk(paths, out_name, *args, **kwargs)
+    res2 = read_zarr(out_name)
+    assert_equal(res1, res2)
+
+
+def test_concatenate_dense(join_type, tmp_path):
```

Comment:
Could this get parameterized on data types? I would like to know this works with CSR arrays as well.

Could you also add more types into the data frame columns and mapping attributes of the AnnDatas? It would be good to be verifying those were behaving as expected.